Basic Pandera Generator for Initial Merge

tfliss opened the pull request:
This is a Pandera generator with basic flat-table functionality. Related to https://github.com/linkml/linkml/issues/671

There are some examples in pandera.rst

To run the tests:

```
poetry install --no-interaction --all-extras
poetry install -E tests
pytest -s -vv --log-cli-level=INFO tests/test_generators/test_panderagen.py
pytest -s -vv --log-cli-level=INFO -m panderagen --with-output tests/test_compliance
```

**Limitations:**

- This is a flat table implementation and class ranges are *not* implemented.
- many other LinkML features are not implemented. The compliance test suite and documentation will be provided.

**Next Steps**

- I will commit the .rst documentation in the next few days with a clear compliance report

**Future Steps**

Another branch/PR (https://github.com/linkml/linkml/pull/2543) has a draft implementation of class ranges and other features, I'll add them back with incremental issues.after the initial review merge.

This is the current compliance summary:

``` csv
name	pandera_polars_class
test_alias	implements
test_enum_alias	implements
test_annotation	implements
test_array	incomplete
test_slot_exactly_one_of	incomplete
test_slot_all_of	incomplete
test_slot_none_of	incomplete
test_cardinality_in_exactly_one_of	incomplete
test_class_any_of	incomplete
test_min_max	incomplete
test_preconditions_combos	incomplete
test_classification_rules	incomplete
test_membership	incomplete
test_attributes	implements
test_type_range	implements
test_min_max_values	implements
test_uri_types	implements
test_date_types	incomplete
test_cardinality	incomplete
test_identifier_is_required	implements
test_non_standard_num_names	incomplete
test_enum	incomplete
test_permissible_value_typing	incomplete
test_enum_hierarchy	incomplete
test_ifabsent	implements
test_import	incomplete
test_basic_class_inheritance	incomplete
test_mixins	incomplete
test_refine_attributes	incomplete
test_slot_usage	incomplete
test_basic_slot_inheritance	incomplete
test_abstract_classes	incomplete
test_instantiates	implements
test_jsonpointer	incomplete
test_descriptions	implements
test_element_uris	implements
test_common_metadata	
test_pattern	incomplete
test_id_prefix	incomplete

```

tfliss commented:
@amc-corey-cox @matentzn @sierra-moxon  I trimmed out many of the features and added a documentation page.
Given the scope I would be surprised if some issues don't come up during review, but I would like to send it out on Monday if possible.

I'll also close the other pandera PR, but I'd like to keep the other branch around for a bit because it has partial implementation of additional featues (I can move it to my fork if that's a better place).

Though it's Friday, I thought I'd give you heads up in the spirit of navigating the PR through the process as smoothly as possible, and also if you have any specific recommendations for reviewers you'd like to have a first look. Have a good weekend!

matentzn commented:
@tfliss sorry I need a bit of handholding for this again. I was able to run the tests and everything looks good.

I would now like to look at my own schemas, so I tried

```
poetry run python -m linkml.generators.panderagen tests/test_generators/input/matrix.yaml
```

But that does not work anymore:

```
No module named linkml.generators.panderagen.__main__; 'linkml.generators.panderagen' is a package and cannot be directly executed
```

What is the right way to generate a pandera schema for testing?

tfliss commented:
@matentzn Thanks for looking at this.  Using the `gen-pandera` command should work now.

`poetry run gen-pandera tests/test_generators/input/matrix.yaml`

matentzn commented:
@tfliss Sorry for coming back so late. When I run the command (I verified the yaml file exists in this location), I get:

```
➜  linkml git:(671-panderagen) ✗ poetry run gen-pandera tests/test_generators/input/matrix.yaml
Usage: gen-pandera [OPTIONS]
Try 'gen-pandera --help' for help.

Error: Got unexpected extra argument (tests/test_generators/input/matrix.yaml)
```

I double checked its installed as well:

```
➜  linkml git:(671-panderagen) ✗ poetry install
Installing dependencies from lock file

No dependencies to install or update

Installing the current project: linkml (0.0.0)
```

When I check the help, I see that:

```
➜  linkml git:(671-panderagen) ✗ poetry run gen-pandera --help                                 
Usage: gen-pandera [OPTIONS]

Options:
  -V, --version         Show the version and exit.
  --template-file TEXT  Optional jinja2 template to use for class generation
  --template-path TEXT  Optional jinja2 template directory within module
  --package TEXT        Package name where relevant for generated class files
  --help                Show this message and exit.
```

When I look at the same for say `gen-project`, I see the `YAMLFILE` (positional argument) in the CLI docs:

```
➜  linkml git:(671-panderagen) ✗ poetry run gen-project --help
Usage: gen-project [OPTIONS] YAMLFILE

  Generate an entire project LinkML schema
```

Can you advice on how to proceed?

tfliss commented:
> Error: Got unexpected extra argument (tests/test_generators/input/matrix.yaml)

Thank you for trying this out @matentzn I have reproduced this message on my side in the debugger and will start on a fix.
Will also resolve the poetry and pyproject conflicts.

tfliss commented:
> linkml git:(671-panderagen) ✗ poetry run gen-pandera tests/test_generators/input/matrix.yaml

@matentzn This command should work now, I had removed the common generator cli parameters which included the yaml file. My latest commit adds that back in and also supports the `linkml generate pandera <yaml>` form of the cli tool. I also added test coverage. Thank you for finding this and please let me know if you uncover any other issues.

tfliss commented:
@sierra-moxon thanks for pointing out the pyproject.toml / poetry.lock conflict this morning, I re-merged from main to resolve it.

sierra-moxon commented:
all tests pass for me, also ran:

 `poetry run gen-pandera tests/test_generators/input/...` with a few different schemas with success. 

thank you @tfliss!!  this is great.  We can pull this in and if we have more iterating to do that is just fine to do in another PR.
